From 704753436afb0dbe1cac4fa53b12871edddaf5be Mon Sep 17 00:00:00 2001
From: Konstantin Klubnichkin <kitsok@yandex-team.ru>
Date: Tue, 8 Nov 2022 18:11:32 +0000
Subject: [PATCH] WIP: Yandex

- Take default config
- Add IBMs config
- Add CONFIG_RSA
- Add ACRY and HACE both to config and DTS
- Fix bootcmd
- Disable raw and legacy image support
- Enable FIT print in SPL
- Enable CONFIG_LOG_MAX_LEVEL=9 for both SPL and main U-Boot
- FIXME Force debug in include/log.h
- Remove debug as it's useless
- Add print FIT in SPL
- Remove SPL GPIO support
- Remove SHA256, SHA384 SPL support
- Remove SPL watchdog support
- Remove more code
- Reduce malloc() heap size
- Refactor ast_sysreset_request() to reduce size
- Print error text on signature verification failure
- Be more verbose in SPL
- Add CONFIG_OF_CONTROL=y
- Reduce logging
- Fix DTS (scu vs. syscon)
- Rework reset via WDT3 (reduce timeout)
- Add SPL board_init into defconfig

%% original patch: 0002-WIP-Yandex.patch

%% original patch: 0002-WIP-Yandex.patch
---
 arch/arm/dts/ast2600.dtsi         | 99 ++++++++++++++++---------------
 arch/arm/dts/yandex-som2600v1.dts | 39 ++----------
 common/image-fit.c                |  4 +-
 common/image-sig.c                |  2 +
 common/spl/spl_ram.c              |  3 +-
 common/spl/spl_spi.c              |  2 +-
 configs/yandex_ast2600_defconfig  | 58 +++++++++++-------
 drivers/sysreset/sysreset_ast.c   | 18 ++----
 include/configs/yandex_ast2600.h  | 21 ++++---
 include/log.h                     |  2 +-
 10 files changed, 120 insertions(+), 128 deletions(-)

diff --git a/arch/arm/dts/ast2600.dtsi b/arch/arm/dts/ast2600.dtsi
index 72f21c834b..ccbc42a1dd 100644
--- a/arch/arm/dts/ast2600.dtsi
+++ b/arch/arm/dts/ast2600.dtsi
@@ -122,7 +122,7 @@
 			compatible = "aspeed,ast2600-fmc";
 			status = "disabled";
 			interrupts = <GIC_SPI 39 IRQ_TYPE_LEVEL_HIGH>;
-			clocks = <&scu ASPEED_CLK_AHB>;
+			clocks = <&syscon ASPEED_CLK_AHB>;
 			num-cs = <3>;
 			flash@0 {
 				reg = < 0 >;
@@ -147,7 +147,7 @@
 			#address-cells = <1>;
 			#size-cells = <0>;
 			compatible = "aspeed,ast2600-spi";
-			clocks = <&scu ASPEED_CLK_AHB>;
+			clocks = <&syscon ASPEED_CLK_AHB>;
 			num-cs = <2>;
 			status = "disabled";
 			flash@0 {
@@ -168,7 +168,7 @@
 			#address-cells = <1>;
 			#size-cells = <0>;
 			compatible = "aspeed,ast2600-spi";
-			clocks = <&scu ASPEED_CLK_AHB>;
+			clocks = <&syscon ASPEED_CLK_AHB>;
 			num-cs = <3>;
 			status = "disabled";
 			flash@0 {
@@ -205,7 +205,7 @@
 			compatible = "aspeed,ast2600-mac", "faraday,ftgmac100";
 			reg = <0x1e660000 0x180>, <0x1e650000 0x4>;
 			interrupts = <GIC_SPI 2 IRQ_TYPE_LEVEL_HIGH>;
-			clocks = <&scu ASPEED_CLK_GATE_MAC1CLK>;
+			clocks = <&syscon ASPEED_CLK_GATE_MAC1CLK>;
 			status = "disabled";
 		};
 
@@ -215,7 +215,7 @@
 			#address-cells = <1>;
 			#size-cells = <0>;
 			interrupts = <GIC_SPI 3 IRQ_TYPE_LEVEL_HIGH>;
-			clocks = <&scu ASPEED_CLK_GATE_MAC2CLK>;
+			clocks = <&syscon ASPEED_CLK_GATE_MAC2CLK>;
 			status = "disabled";
 		};
 
@@ -225,7 +225,7 @@
 			#address-cells = <1>;
 			#size-cells = <0>;
 			interrupts = <GIC_SPI 32 IRQ_TYPE_LEVEL_HIGH>;
-			clocks = <&scu ASPEED_CLK_GATE_MAC3CLK>;
+			clocks = <&syscon ASPEED_CLK_GATE_MAC3CLK>;
 			status = "disabled";
 		};
 
@@ -235,7 +235,7 @@
 			#address-cells = <1>;
 			#size-cells = <0>;
 			interrupts = <GIC_SPI 33 IRQ_TYPE_LEVEL_HIGH>;
-			clocks = <&scu ASPEED_CLK_GATE_MAC4CLK>;
+			clocks = <&syscon ASPEED_CLK_GATE_MAC4CLK>;
 			status = "disabled";
 		};
 
@@ -254,7 +254,7 @@
 			compatible = "aspeed,aspeed-ehci", "usb-ehci";
 			reg = <0x1e6a1000 0x100>;
 			interrupts = <GIC_SPI 5 IRQ_TYPE_LEVEL_HIGH>;
-			clocks = <&scu ASPEED_CLK_GATE_USBPORT1CLK>;
+			clocks = <&syscon ASPEED_CLK_GATE_USBPORT1CLK>;
 			pinctrl-names = "default";
 			pinctrl-0 = <&pinctrl_usb2ah_default>;
 			status = "disabled";
@@ -264,7 +264,7 @@
 			compatible = "aspeed,aspeed-ehci", "usb-ehci";
 			reg = <0x1e6a3000 0x100>;
 			interrupts = <GIC_SPI 9 IRQ_TYPE_LEVEL_HIGH>;
-			clocks = <&scu ASPEED_CLK_GATE_USBPORT2CLK>;
+			clocks = <&syscon ASPEED_CLK_GATE_USBPORT2CLK>;
 			pinctrl-names = "default";
 			pinctrl-0 = <&pinctrl_usb2bh_default>;
 			status = "disabled";
@@ -278,13 +278,13 @@
 			u-boot,dm-pre-reloc;
 
 			syscon: syscon@1e6e2000 {
-				compatible = "aspeed,g6-scu", "syscon", "simple-mfd";
+				compatible = "aspeed,g6-scu", "syscon", "simple-mfd";
 				reg = <0x1e6e2000 0x1000>;
+				ranges = <0 0x1e6e2000 0x1000>;
 				#address-cells = <1>;
 				#size-cells = <1>;
 				#clock-cells = <1>;
 				#reset-cells = <1>;
-				ranges = <0 0x1e6e2000 0x1000>;
 				u-boot,dm-pre-reloc;
 
 				pinctrl: pinctrl {
@@ -321,8 +321,9 @@
 				compatible = "aspeed,ast2600-hace";
 				reg = <0x1e6d0000 0x200>;
 				interrupts = <GIC_SPI 4 IRQ_TYPE_LEVEL_HIGH>;
-				clocks = <&scu ASPEED_CLK_GATE_YCLK>;
+				clocks = <&syscon ASPEED_CLK_GATE_YCLK>;
 				clock-names = "yclk";
+				resets = <&syscon ASPEED_RESET_HACE>;
 				status = "disabled";
 			};
 
@@ -331,7 +332,7 @@
 				reg = <0x1e6fa000 0x1000>,
 				      <0x1e710000 0x20000>;
 				interrupts = <GIC_SPI 160 IRQ_TYPE_LEVEL_HIGH>;
-				clocks = <&scu ASPEED_CLK_GATE_RSAECCCLK>;
+				clocks = <&syscon ASPEED_CLK_GATE_RSAECCCLK>;
 				clock-names = "rsaeccclk";
 				status = "disabled";
 			};
@@ -353,7 +354,7 @@
 				reg = <0x1e740000 0x1000>;
 				interrupts = <GIC_SPI 43 IRQ_TYPE_LEVEL_HIGH>;
 				interrupt-controller;
-				clocks = <&scu ASPEED_CLK_GATE_SDCLK>, <&scu ASPEED_CLK_GATE_SDEXTCLK>;
+				clocks = <&syscon ASPEED_CLK_GATE_SDCLK>, <&syscon ASPEED_CLK_GATE_SDEXTCLK>;
 				clock-names = "ctrlclk", "extclk";
 				#address-cells = <1>;
 				#size-cells = <1>;
@@ -365,7 +366,7 @@
 					interrupts = <0>;
 					interrupt-parent = <&sdhci>;
 					sdhci,auto-cmd12;
-					clocks = <&scu ASPEED_CLK_SDIO>;
+					clocks = <&syscon ASPEED_CLK_SDIO>;
 					status = "disabled";
 				};
 
@@ -375,7 +376,7 @@
 					interrupts = <1>;
 					interrupt-parent = <&sdhci>;
 					sdhci,auto-cmd12;
-					clocks = <&scu ASPEED_CLK_SDIO>;
+					clocks = <&syscon ASPEED_CLK_SDIO>;
 					status = "disabled";
 				};
 			};
@@ -386,7 +387,7 @@
 				reg = <0x1e750000 0x1000>;
 				interrupts = <GIC_SPI 15 IRQ_TYPE_LEVEL_HIGH>;
 				interrupt-controller;
-				clocks = <&scu ASPEED_CLK_GATE_EMMCCLK>, <&scu ASPEED_CLK_GATE_EMMCEXTCLK>;
+				clocks = <&syscon ASPEED_CLK_GATE_EMMCCLK>, <&syscon ASPEED_CLK_GATE_EMMCEXTCLK>;
 				clock-names = "ctrlclk", "extclk";
 				#address-cells = <1>;
 				#size-cells = <1>;
@@ -397,7 +398,7 @@
 					reg = <0x100 0x100>;
 					interrupts = <0>;
 					interrupt-parent = <&emmc>;
-					clocks = <&scu ASPEED_CLK_EMMC>;
+					clocks = <&syscon ASPEED_CLK_EMMC>;
 					status = "disabled";
 				};
 			};
@@ -462,7 +463,7 @@
 				reg = <0x1e783000 0x20>;
 				reg-shift = <2>;
 				interrupts = <GIC_SPI 47 IRQ_TYPE_LEVEL_HIGH>;
-				clocks = <&scu ASPEED_CLK_GATE_UART1CLK>;
+				clocks = <&syscon ASPEED_CLK_GATE_UART1CLK>;
 				clock-frequency = <1846154>;
 				no-loopback-test;
 				u-boot,dm-pre-reloc;
@@ -476,7 +477,7 @@
 				reg = <0x1e784000 0x1000>;
 				reg-shift = <2>;
 				interrupts = <GIC_SPI 36 IRQ_TYPE_LEVEL_HIGH>;
-				clocks = <&scu ASPEED_CLK_GATE_UART5CLK>;
+				clocks = <&syscon ASPEED_CLK_GATE_UART5CLK>;
 				clock-frequency = <1846154>;
 				no-loopback-test;
 				u-boot,dm-pre-reloc;
@@ -617,7 +618,7 @@
 				reg = <0x1e78d000 0x20>;
 				reg-shift = <2>;
 				interrupts = <GIC_SPI 48 IRQ_TYPE_LEVEL_HIGH>;
-				clocks = <&scu ASPEED_CLK_GATE_UART2CLK>;
+				clocks = <&syscon ASPEED_CLK_GATE_UART2CLK>;
 				clock-frequency = <1846154>;
 				no-loopback-test;
 				pinctrl-names = "default";
@@ -631,7 +632,7 @@
 				reg = <0x1e78e000 0x20>;
 				reg-shift = <2>;
 				interrupts = <GIC_SPI 49 IRQ_TYPE_LEVEL_HIGH>;
-				clocks = <&scu ASPEED_CLK_GATE_UART3CLK>;
+				clocks = <&syscon ASPEED_CLK_GATE_UART3CLK>;
 				clock-frequency = <1846154>;
 				no-loopback-test;
 				pinctrl-names = "default";
@@ -645,7 +646,7 @@
 				reg = <0x1e78f000 0x20>;
 				reg-shift = <2>;
 				interrupts = <GIC_SPI 50 IRQ_TYPE_LEVEL_HIGH>;
-				clocks = <&scu ASPEED_CLK_GATE_UART4CLK>;
+				clocks = <&syscon ASPEED_CLK_GATE_UART4CLK>;
 				clock-frequency = <1846154>;
 				no-loopback-test;
 				u-boot,dm-pre-reloc;
@@ -667,7 +668,7 @@
 				interrupts = <GIC_SPI 100 IRQ_TYPE_LEVEL_HIGH>;
 				pinctrl-names = "default";
 				pinctrl-0 = <&pinctrl_fsi1_default>;
-				clocks = <&scu ASPEED_CLK_GATE_FSICLK>;
+				clocks = <&syscon ASPEED_CLK_GATE_FSICLK>;
 				status = "disabled";
 			};
 
@@ -677,7 +678,7 @@
 				interrupts = <GIC_SPI 101 IRQ_TYPE_LEVEL_HIGH>;
 				pinctrl-names = "default";
 				pinctrl-0 = <&pinctrl_fsi2_default>;
-				clocks = <&scu ASPEED_CLK_GATE_FSICLK>;
+				clocks = <&syscon ASPEED_CLK_GATE_FSICLK>;
 				status = "disabled";
 			};
 
@@ -686,7 +687,7 @@
 				reg = <0x1e790000 0x20>;
 				reg-shift = <2>;
 				interrupts = <GIC_SPI 57 IRQ_TYPE_LEVEL_HIGH>;
-				clocks = <&scu ASPEED_CLK_GATE_UART6CLK>;
+				clocks = <&syscon ASPEED_CLK_GATE_UART6CLK>;
 				clock-frequency = <1846154>;
 				no-loopback-test;
 				status = "disabled";
@@ -697,7 +698,7 @@
 				reg = <0x1e790100 0x20>;
 				reg-shift = <2>;
 				interrupts = <GIC_SPI 58 IRQ_TYPE_LEVEL_HIGH>;
-				clocks = <&scu ASPEED_CLK_GATE_UART7CLK>;
+				clocks = <&syscon ASPEED_CLK_GATE_UART7CLK>;
 				clock-frequency = <1846154>;
 				no-loopback-test;
 				status = "disabled";
@@ -708,7 +709,7 @@
 				reg = <0x1e790200 0x20>;
 				reg-shift = <2>;
 				interrupts = <GIC_SPI 59 IRQ_TYPE_LEVEL_HIGH>;
-				clocks = <&scu ASPEED_CLK_GATE_UART8CLK>;
+				clocks = <&syscon ASPEED_CLK_GATE_UART8CLK>;
 				clock-frequency = <1846154>;
 				no-loopback-test;
 				status = "disabled";
@@ -719,7 +720,7 @@
 				reg = <0x1e790300 0x20>;
 				reg-shift = <2>;
 				interrupts = <GIC_SPI 60 IRQ_TYPE_LEVEL_HIGH>;
-				clocks = <&scu ASPEED_CLK_GATE_UART9CLK>;
+				clocks = <&syscon ASPEED_CLK_GATE_UART9CLK>;
 				clock-frequency = <1846154>;
 				no-loopback-test;
 				status = "disabled";
@@ -730,7 +731,7 @@
 				reg = <0x1e790400 0x20>;
 				reg-shift = <2>;
 				interrupts = <GIC_SPI 61 IRQ_TYPE_LEVEL_HIGH>;
-				clocks = <&scu ASPEED_CLK_GATE_UART10CLK>;
+				clocks = <&syscon ASPEED_CLK_GATE_UART10CLK>;
 				clock-frequency = <1846154>;
 				no-loopback-test;
 				status = "disabled";
@@ -741,7 +742,7 @@
 				reg = <0x1e790400 0x20>;
 				reg-shift = <2>;
 				interrupts = <GIC_SPI 62 IRQ_TYPE_LEVEL_HIGH>;
-				clocks = <&scu ASPEED_CLK_GATE_UART11CLK>;
+				clocks = <&syscon ASPEED_CLK_GATE_UART11CLK>;
 				clock-frequency = <1846154>;
 				no-loopback-test;
 				status = "disabled";
@@ -752,7 +753,7 @@
 				reg = <0x1e790600 0x20>;
 				reg-shift = <2>;
 				interrupts = <GIC_SPI 63 IRQ_TYPE_LEVEL_HIGH>;
-				clocks = <&scu ASPEED_CLK_GATE_UART12CLK>;
+				clocks = <&syscon ASPEED_CLK_GATE_UART12CLK>;
 				clock-frequency = <1846154>;
 				no-loopback-test;
 				status = "disabled";
@@ -763,7 +764,7 @@
 				reg = <0x1e790700 0x20>;
 				reg-shift = <2>;
 				interrupts = <GIC_SPI 64 IRQ_TYPE_LEVEL_HIGH>;
-				clocks = <&scu ASPEED_CLK_GATE_UART13CLK>;
+				clocks = <&syscon ASPEED_CLK_GATE_UART13CLK>;
 				clock-frequency = <1846154>;
 				no-loopback-test;
 				status = "disabled";
@@ -799,7 +800,7 @@
 		compatible = "aspeed,ast2600-i2c-bus";
 		bus-frequency = <100000>;
 		interrupts = <GIC_SPI 110 IRQ_TYPE_LEVEL_HIGH>;
-		clocks = <&scu ASPEED_CLK_APB2>;
+		clocks = <&syscon ASPEED_CLK_APB2>;
 		pinctrl-names = "default";
 		pinctrl-0 = <&pinctrl_i2c1_default>;
 		status = "disabled";
@@ -814,7 +815,7 @@
 		compatible = "aspeed,ast2600-i2c-bus";
 		bus-frequency = <100000>;
 		interrupts = <GIC_SPI 111 IRQ_TYPE_LEVEL_HIGH>;
-		clocks = <&scu ASPEED_CLK_APB2>;
+		clocks = <&syscon ASPEED_CLK_APB2>;
 		pinctrl-names = "default";
 		pinctrl-0 = <&pinctrl_i2c2_default>;
 		status = "disabled";
@@ -829,7 +830,7 @@
 		compatible = "aspeed,ast2600-i2c-bus";
 		bus-frequency = <100000>;
 		interrupts = <GIC_SPI 112 IRQ_TYPE_LEVEL_HIGH>;
-		clocks = <&scu ASPEED_CLK_APB2>;
+		clocks = <&syscon ASPEED_CLK_APB2>;
 		pinctrl-names = "default";
 		pinctrl-0 = <&pinctrl_i2c3_default>;
 	};
@@ -843,7 +844,7 @@
 		compatible = "aspeed,ast2600-i2c-bus";
 		bus-frequency = <100000>;
 		interrupts = <GIC_SPI 113 IRQ_TYPE_LEVEL_HIGH>;
-		clocks = <&scu ASPEED_CLK_APB2>;
+		clocks = <&syscon ASPEED_CLK_APB2>;
 		pinctrl-names = "default";
 		pinctrl-0 = <&pinctrl_i2c4_default>;
 	};
@@ -857,7 +858,7 @@
 		compatible = "aspeed,ast2600-i2c-bus";
 		bus-frequency = <100000>;
 		interrupts = <GIC_SPI 114 IRQ_TYPE_LEVEL_HIGH>;
-		clocks = <&scu ASPEED_CLK_APB2>;
+		clocks = <&syscon ASPEED_CLK_APB2>;
 		pinctrl-names = "default";
 		pinctrl-0 = <&pinctrl_i2c5_default>;
 	};
@@ -871,7 +872,7 @@
 		compatible = "aspeed,ast2600-i2c-bus";
 		bus-frequency = <100000>;
 		interrupts = <GIC_SPI 115 IRQ_TYPE_LEVEL_HIGH>;
-		clocks = <&scu ASPEED_CLK_APB2>;
+		clocks = <&syscon ASPEED_CLK_APB2>;
 		pinctrl-names = "default";
 		pinctrl-0 = <&pinctrl_i2c6_default>;
 	};
@@ -885,7 +886,7 @@
 		compatible = "aspeed,ast2600-i2c-bus";
 		bus-frequency = <100000>;
 		interrupts = <GIC_SPI 116 IRQ_TYPE_LEVEL_HIGH>;
-		clocks = <&scu ASPEED_CLK_APB2>;
+		clocks = <&syscon ASPEED_CLK_APB2>;
 		pinctrl-names = "default";
 		pinctrl-0 = <&pinctrl_i2c7_default>;
 	};
@@ -899,7 +900,7 @@
 		compatible = "aspeed,ast2600-i2c-bus";
 		bus-frequency = <100000>;
 		interrupts = <GIC_SPI 117 IRQ_TYPE_LEVEL_HIGH>;
-		clocks = <&scu ASPEED_CLK_APB2>;
+		clocks = <&syscon ASPEED_CLK_APB2>;
 		pinctrl-names = "default";
 		pinctrl-0 = <&pinctrl_i2c8_default>;
 	};
@@ -913,7 +914,7 @@
 		compatible = "aspeed,ast2600-i2c-bus";
 		bus-frequency = <100000>;
 		interrupts = <GIC_SPI 118 IRQ_TYPE_LEVEL_HIGH>;
-		clocks = <&scu ASPEED_CLK_APB2>;
+		clocks = <&syscon ASPEED_CLK_APB2>;
 		pinctrl-names = "default";
 		pinctrl-0 = <&pinctrl_i2c9_default>;
 	};
@@ -927,7 +928,7 @@
 		compatible = "aspeed,ast2600-i2c-bus";
 		bus-frequency = <100000>;
 		interrupts = <GIC_SPI 119 IRQ_TYPE_LEVEL_HIGH>;
-		clocks = <&scu ASPEED_CLK_APB2>;
+		clocks = <&syscon ASPEED_CLK_APB2>;
 		pinctrl-names = "default";
 		pinctrl-0 = <&pinctrl_i2c10_default>;
 		status = "disabled";
@@ -942,7 +943,7 @@
 		compatible = "aspeed,ast2600-i2c-bus";
 		bus-frequency = <100000>;
 		interrupts = <GIC_SPI 120 IRQ_TYPE_LEVEL_HIGH>;
-		clocks = <&scu ASPEED_CLK_APB2>;
+		clocks = <&syscon ASPEED_CLK_APB2>;
 		pinctrl-names = "default";
 		pinctrl-0 = <&pinctrl_i2c11_default>;
 		status = "disabled";
@@ -957,7 +958,7 @@
 		compatible = "aspeed,ast2600-i2c-bus";
 		bus-frequency = <100000>;
 		interrupts = <GIC_SPI 121 IRQ_TYPE_LEVEL_HIGH>;
-		clocks = <&scu ASPEED_CLK_APB2>;
+		clocks = <&syscon ASPEED_CLK_APB2>;
 		pinctrl-names = "default";
 		pinctrl-0 = <&pinctrl_i2c12_default>;
 		status = "disabled";
@@ -972,7 +973,7 @@
 		compatible = "aspeed,ast2600-i2c-bus";
 		bus-frequency = <100000>;
 		interrupts = <GIC_SPI 122 IRQ_TYPE_LEVEL_HIGH>;
-		clocks = <&scu ASPEED_CLK_APB2>;
+		clocks = <&syscon ASPEED_CLK_APB2>;
 		pinctrl-names = "default";
 		pinctrl-0 = <&pinctrl_i2c13_default>;
 		status = "disabled";
@@ -987,7 +988,7 @@
 		compatible = "aspeed,ast2600-i2c-bus";
 		bus-frequency = <100000>;
 		interrupts = <GIC_SPI 123 IRQ_TYPE_LEVEL_HIGH>;
-		clocks = <&scu ASPEED_CLK_APB2>;
+		clocks = <&syscon ASPEED_CLK_APB2>;
 		pinctrl-names = "default";
 		pinctrl-0 = <&pinctrl_i2c14_default>;
 		status = "disabled";
@@ -1002,7 +1003,7 @@
 		compatible = "aspeed,ast2600-i2c-bus";
 		bus-frequency = <100000>;
 		interrupts = <GIC_SPI 124 IRQ_TYPE_LEVEL_HIGH>;
-		clocks = <&scu ASPEED_CLK_APB2>;
+		clocks = <&syscon ASPEED_CLK_APB2>;
 		pinctrl-names = "default";
 		pinctrl-0 = <&pinctrl_i2c15_default>;
 		status = "disabled";
@@ -1017,7 +1018,7 @@
 		compatible = "aspeed,ast2600-i2c-bus";
 		bus-frequency = <100000>;
 		interrupts = <GIC_SPI 125 IRQ_TYPE_LEVEL_HIGH>;
-		clocks = <&scu ASPEED_CLK_APB2>;
+		clocks = <&syscon ASPEED_CLK_APB2>;
 		pinctrl-names = "default";
 		pinctrl-0 = <&pinctrl_i2c16_default>;
 		status = "disabled";
diff --git a/arch/arm/dts/yandex-som2600v1.dts b/arch/arm/dts/yandex-som2600v1.dts
index 587b8e1375..e01be14dc0 100644
--- a/arch/arm/dts/yandex-som2600v1.dts
+++ b/arch/arm/dts/yandex-som2600v1.dts
@@ -201,44 +201,10 @@
 	};
 };
 
-&emmc {
-	u-boot,dm-pre-reloc;
-	timing-phase = <0x700ff>;
-};
-
-&emmc_slot0 {
-	u-boot,dm-pre-reloc;
-	status = "okay";
-	bus-width = <4>;
-	pinctrl-names = "default";
-	pinctrl-0 = <&pinctrl_emmc_default>;
-	sdhci-drive-type = <1>;
-};
-
 &sdhci {
 	timing-phase = <0xc6ffff>;
 };
 
-&sdhci_slot0 {
-	status = "okay";
-	bus-width = <4>;
-	pwr-gpios = <&gpio0 ASPEED_GPIO(V, 0) GPIO_ACTIVE_HIGH>;
-	pwr-sw-gpios = <&gpio0 ASPEED_GPIO(V, 1) GPIO_ACTIVE_HIGH>;
-	pinctrl-names = "default";
-	pinctrl-0 = <&pinctrl_sd1_default>;
-	sdhci-drive-type = <1>;
-};
-
-&sdhci_slot1 {
-	status = "okay";
-	bus-width = <4>;
-	pwr-gpios = <&gpio0 ASPEED_GPIO(V, 2) GPIO_ACTIVE_HIGH>;
-	pwr-sw-gpios = <&gpio0 ASPEED_GPIO(V, 3) GPIO_ACTIVE_HIGH>;
-	pinctrl-names = "default";
-	pinctrl-0 = <&pinctrl_sd2_default>;
-	sdhci-drive-type = <1>;
-};
-
 &i2c4 {
 	status = "okay";
 
@@ -314,6 +280,11 @@
 	status = "okay";
 };
 
+&acry {
+	u-boot,dm-pre-reloc;
+	status = "okay";
+};
+
 &gpio0 {
 	u-boot,dm-pre-reloc;
 };
diff --git a/common/image-fit.c b/common/image-fit.c
index 3c8667f93d..70b4965402 100644
--- a/common/image-fit.c
+++ b/common/image-fit.c
@@ -1242,8 +1242,10 @@ int fit_image_verify_with_data(const void *fit, int image_noffset,
 			 * an image validation failure. See the call to
 			 * fit_image_verify_required_sigs() above.
 			 */
-			if (ret)
+			if (ret) {
 				puts("- ");
+				printf("%s ", err_msg);
+			}
 			else
 				puts("+ ");
 		}
diff --git a/common/image-sig.c b/common/image-sig.c
index c94854ef8b..91e1034472 100644
--- a/common/image-sig.c
+++ b/common/image-sig.c
@@ -299,6 +299,7 @@ static int fit_image_verify_sig(const void *fit, int image_noffset,
 							size, -1, &err_msg);
 			if (ret) {
 				puts("- ");
+				printf("%s ", err_msg);
 			} else {
 				puts("+ ");
 				verified = 1;
@@ -518,6 +519,7 @@ static int fit_config_verify_sig(const void *fit, int conf_noffset,
 						   conf_noffset, &err_msg);
 			if (ret) {
 				puts("- ");
+				printf("%s ", err_msg);
 			} else {
 				puts("+ ");
 				verified = 1;
diff --git a/common/spl/spl_ram.c b/common/spl/spl_ram.c
index 954e91a004..5d88378838 100644
--- a/common/spl/spl_ram.c
+++ b/common/spl/spl_ram.c
@@ -44,7 +44,8 @@ static int spl_ram_load_image(struct spl_image_info *spl_image,
 	    image_get_magic(header) == FDT_MAGIC) {
 		struct spl_load_info load;
 
-		debug("Found FIT\n");
+		debug("Found FIT in RAM\n");
+		fit_print_contents(header);
 		load.bl_len = 1;
 		load.read = spl_ram_load_read;
 		spl_load_simple_fit(spl_image, &load, 0, header);
diff --git a/common/spl/spl_spi.c b/common/spl/spl_spi.c
index 8cd4830a39..b2f1f0d468 100644
--- a/common/spl/spl_spi.c
+++ b/common/spl/spl_spi.c
@@ -122,7 +122,7 @@ static int spl_spi_load_image(struct spl_image_info *spl_image,
 			   image_get_magic(header) == FDT_MAGIC) {
 			struct spl_load_info load;
 
-			debug("Found FIT\n");
+			// debug("Found FIT on SPI\n");
 			load.dev = flash;
 			load.priv = NULL;
 			load.filename = NULL;
diff --git a/configs/yandex_ast2600_defconfig b/configs/yandex_ast2600_defconfig
index 2fc5144cc4..9d642b2537 100644
--- a/configs/yandex_ast2600_defconfig
+++ b/configs/yandex_ast2600_defconfig
@@ -12,7 +12,6 @@ CONFIG_TARGET_YANDEX_AST2600=y
 CONFIG_SPL_LIBCOMMON_SUPPORT=y
 CONFIG_SPL_LIBGENERIC_SUPPORT=y
 CONFIG_SYS_MALLOC_F_LEN=0x2000
-CONFIG_SPL_MMC_SUPPORT=y
 CONFIG_SPL_SERIAL_SUPPORT=y
 CONFIG_SPL_DRIVERS_MISC_SUPPORT=y
 CONFIG_ENV_SIZE=0x10000
@@ -24,23 +23,20 @@ CONFIG_ARMV7_PSCI_NR_CPUS=2
 CONFIG_NR_DRAM_BANKS=1
 CONFIG_FIT=y
 CONFIG_USE_BOOTARGS=y
-CONFIG_BOOTARGS="console=ttyS4,115200n8 root=/dev/ram rw"
+CONFIG_BOOTARGS="console=ttyS4,115200n8 root=/dev/ram rw earlyprintk"
 CONFIG_USE_BOOTCOMMAND=y
-CONFIG_BOOTCOMMAND="bootm 20100000"
+CONFIG_BOOTCOMMAND="run bootspi"
 CONFIG_SYS_CONSOLE_ENV_OVERWRITE=y
 CONFIG_DISPLAY_BOARDINFO_LATE=y
 CONFIG_ARCH_EARLY_INIT_R=y
 CONFIG_BOARD_EARLY_INIT_F=y
-CONFIG_SPL_RAW_IMAGE_SUPPORT=y
-CONFIG_SPL_LEGACY_IMAGE_SUPPORT=y
 CONFIG_SPL_SYS_MALLOC_SIMPLE=y
 CONFIG_SPL_STACK_R=y
 CONFIG_SPL_SEPARATE_BSS=y
 CONFIG_SPL_DM_RESET=y
 CONFIG_SPL_RAM_SUPPORT=y
 CONFIG_SPL_RAM_DEVICE=y
-CONFIG_SPL_WATCHDOG_SUPPORT=y
-CONFIG_SPL_YMODEM_SUPPORT=y
+# REMOVE CONFIG_SPL_WATCHDOG_SUPPORT=y
 CONFIG_HUSH_PARSER=y
 # CONFIG_AUTO_COMPLETE is not set
 CONFIG_SYS_PROMPT="ast# "
@@ -55,7 +51,7 @@ CONFIG_CMD_GPIO=y
 CONFIG_CMD_GPT=y
 # CONFIG_RANDOM_UUID is not set
 CONFIG_CMD_I2C=y
-CONFIG_CMD_MMC=y
+# CONFIG_CMD_MMC=y
 CONFIG_CMD_PART=y
 CONFIG_CMD_SF=y
 CONFIG_CMD_DHCP=y
@@ -89,10 +85,9 @@ CONFIG_DM_I2C=y
 CONFIG_SYS_I2C_ASPEED=y
 CONFIG_MISC=y
 CONFIG_ASPEED_AHBC=y
-CONFIG_DM_MMC=y
-CONFIG_SPL_MMC_TINY=y
-CONFIG_MMC_SDHCI=y
-CONFIG_MMC_SDHCI_ASPEED=y
+# CONFIG_DM_MMC=y
+# CONFIG_MMC_SDHCI=y
+# CONFIG_MMC_SDHCI_ASPEED=y
 CONFIG_DM_SPI_FLASH=y
 CONFIG_SPI_FLASH=y
 CONFIG_SPI_FLASH_GIGADEVICE=y
@@ -118,7 +113,6 @@ CONFIG_SYSRESET=y
 CONFIG_WDT=y
 CONFIG_USE_TINY_PRINTF=y
 CONFIG_TPM=y
-CONFIG_SPL_TPM=y
 # CONFIG_EFI_LOADER is not set
 CONFIG_BOARD_LATE_INIT=y
 CONFIG_ASPEED_BYPASS_SELFTEST=y
@@ -133,21 +127,45 @@ CONFIG_PHY_NCSI=n
 # IBM P10BMC
 CONFIG_ASPEED_HACE_V1=y
 CONFIG_ENV_OFFSET=0xE0000
-CONFIG_FIT_ENABLE_SHA384_SUPPORT=y
+# FIXME
+CONFIG_FIT_ENABLE_SHA384_SUPPORT=n
 CONFIG_FIT_ENABLE_SHA512_SUPPORT=y
 CONFIG_FIT_SIGNATURE=y
+# Removal causes fail of HACE engine
 CONFIG_SPL_BOARD_INIT=y
 CONFIG_SPL_FIT_IMAGE_TINY=y
 CONFIG_SPL_FIT_SIGNATURE=y
-CONFIG_SPL_GPIO_HOG=y
-CONFIG_SPL_GPIO_SUPPORT=y
+# REMOVE CONFIG_SPL_GPIO_HOG=y
+# REMOVE CONFIG_SPL_GPIO_SUPPORT=y
 CONFIG_SPL_LOAD_FIT=y
-CONFIG_SPL_SHA1_SUPPORT=y
-CONFIG_SPL_SHA256_SUPPORT=y
-CONFIG_SPL_SHA384_SUPPORT=y
+# FIXME
+CONFIG_SPL_SHA1_SUPPORT=n
+CONFIG_SPL_SHA256_SUPPORT=n
+CONFIG_SPL_SHA384_SUPPORT=n
 CONFIG_SPL_SHA512_SUPPORT=y
 CONFIG_SPL_STACK_R_ADDR=0x90300000
+# FIXME WARNING
+# CONFIG_SPL_SYS_MALLOC_F_LEN=0x800
 CONFIG_SPL_SYS_MALLOC_F_LEN=0x1000
 CONFIG_SPL_TINY_MEMSET=y
 
-CONFIG_ASPEED_ECC=y
+CONFIG_RSA=y
+CONFIG_ASPEED_HACE=y
+CONFIG_ASPEED_ACRY=y
+# CONFIG_SPL_RAW_IMAGE_SUPPORT=y
+# CONFIG_SPL_LEGACY_IMAGE_SUPPORT=y
+CONFIG_SPL_CRYPTO_SUPPORT=y
+CONFIG_SPL_FIT_PRINT=y
+
+# Debug
+# CONFIG_DEBUG_UART=y
+# CONFIG_DEBUG_UART_BASE=0x1e784000
+# CONFIG_DEBUG_UART_CLOCK=1846154
+# CONFIG_DEBUG_UART_SHIFT=2
+# CONFIG_LOG_MAX_LEVEL=9
+# CONFIG_SPL_LOG_MAX_LEVEL=9
+
+# After reading https://www.timesys.com/security/securing-u-boot-a-guide-to-mitigating-common-attack-vectors/
+CONFIG_FIT_VERBOSE=y
+CONFIG_OF_CONTROL=y
+CONFIG_SHA_PROG_HW_ACCEL=y
diff --git a/drivers/sysreset/sysreset_ast.c b/drivers/sysreset/sysreset_ast.c
index d5d8736dc6..37265a520b 100644
--- a/drivers/sysreset/sysreset_ast.c
+++ b/drivers/sysreset/sysreset_ast.c
@@ -9,23 +9,17 @@
 #include <dm.h>
 #include <errno.h>
 #include <sysreset.h>
-#include <wdt.h>
 #include <asm/io.h>
 #include <linux/err.h>
 
+#define AST_WDT3_BASE 0x1E785080
+#define WDT3_TIMEOUT 50000
 static int ast_sysreset_request(struct udevice *dev, enum sysreset_t type)
 {
-	struct udevice *wdt;
-	int ret = uclass_first_device(UCLASS_WDT, &wdt);
-
-	if (ret)
-		return ret;
-
-	ret = wdt_expire_now(wdt, 0);
-	if (ret) {
-		debug("Sysreset failed: %d", ret);
-		return ret;
-	}
+	printf("WDT3 enabled, reset now\n");
+	writel((u32) WDT3_TIMEOUT, AST_WDT3_BASE + 0x04);
+	writel(0x4755, AST_WDT3_BASE + 0x08);
+	writel(0x13, AST_WDT3_BASE + 0x0c);
 
 	return -EINPROGRESS;
 }
diff --git a/include/configs/yandex_ast2600.h b/include/configs/yandex_ast2600.h
index f4f1b46d89..101566bdde 100644
--- a/include/configs/yandex_ast2600.h
+++ b/include/configs/yandex_ast2600.h
@@ -17,20 +17,23 @@
 /* Memory Info */
 #define CONFIG_SYS_LOAD_ADDR		0x83000000
 
-#ifdef CONFIG_SPL_TINY
-#ifdef CONFIG_SPL_BUILD
-#define CONFIG_SYS_NS16550_REG_SIZE 2
-#endif
-#endif
-
 /* SPL */
 #define CONFIG_SPL_TEXT_BASE		0x00000000
 #define CONFIG_SPL_MAX_SIZE		0x00010000
 #define CONFIG_SPL_STACK		0x10016000
 #define CONFIG_SPL_BSS_START_ADDR	0x90000000
 #define CONFIG_SPL_BSS_MAX_SIZE		0x00100000
-#define CONFIG_SPL_LOAD_FIT_ADDRESS     0x00010000
-
-#define CONFIG_SUPPORT_EMMC_BOOT
+#define CONFIG_SPL_LOAD_FIT_ADDRESS	0x00010000
+
+/* Extra ENV for Boot Command */
+#define STR_HELPER(n)	#n
+#define STR(n)		STR_HELPER(n)
+
+#undef CONFIG_EXTRA_ENV_SETTINGS
+#define CONFIG_EXTRA_ENV_SETTINGS	\
+	"loadaddr=" STR(CONFIG_SYS_LOAD_ADDR) "\0"	\
+	"bootspi=fdt addr 20100000 && fdt header get fitsize totalsize && cp.b 20100000 ${loadaddr} ${fitsize} && bootm; echo Error loading kernel FIT image\0"	\
+	"verify=yes\0"	\
+	""
 
 #endif	/* __CONFIG_H */
diff --git a/include/log.h b/include/log.h
index 7566ba7f2d..73e7aa5548 100644
--- a/include/log.h
+++ b/include/log.h
@@ -158,7 +158,7 @@ int _log(enum log_category_t cat, enum log_level_t level, const char *file,
  */
 #define debug_cond(cond, fmt, args...)			\
 	do {						\
-		if (cond)				\
+		if (cond)                               \
 			printf(pr_fmt(fmt), ##args);	\
 	} while (0)
 
-- 
2.36.0

